adapters/base/src/main/java/org/apache/cassandra/sidecar/adapters/base/TokenRangeReplicaProvider.java

```diff
@@ -0,0 +1,263 @@
+    {
+        return replicaMappings.entrySet()
+                              .stream()
+                              .map(entry -> generateTokenRangeReplicas(new BigInteger(entry.getKey().get(0)),
+                                                                       new BigInteger(entry.getKey().get(1)),
+                                                                       partitioner,
+                                                                       new HashSet<>(entry.getValue())))
+                              .flatMap(Collection::stream)
+                              .collect(toList());
+    }
+
+
+    private List<ReplicaInfo> readReplicasFromReplicaMapping(List<TokenRangeReplicas> naturalTokenRangeReplicas,
+                                                             Map<String, String> hostToDatacenter)
+    {
+        Map<String, String> hostToDatacenter2 = buildHostToDatacenterMapping(naturalTokenRangeReplicas);
+        return naturalTokenRangeReplicas.stream()
+                                        .sorted()
+                                        .map(rep -> {
+                                            Map<String, List<String>> replicasByDc
+                                            = replicasByDataCenter(hostToDatacenter2, rep.replicaSet());
+
+                                            return new ReplicaInfo(rep.start().toString(),
+                                                                   rep.end().toString(),
+                                                                   replicasByDc);
+                                        })
+                                        .collect(toList());
+    }
+
+    private Map<String, String> buildHostToDatacenterMapping(List<TokenRangeReplicas> replicaSet)
+    {
+        EndpointSnitchJmxOperations endpointSnitchInfo = initializeEndpointProxy();
+
+        return replicaSet.stream()
+                         .map(TokenRangeReplicas::replicaSet)
+                         .flatMap(Collection::stream)
+                         .distinct()
+                         .collect(Collectors.toMap(Function.identity(),
+                                                   (String host) -> getDatacenter(endpointSnitchInfo, host)));
+    }
+
+    private String getDatacenter(EndpointSnitchJmxOperations endpointSnitchInfo, String host)
+    {
+        try
+        {
+            return endpointSnitchInfo.getDatacenter(host);
+        }
+        catch (UnknownHostException e)
+        {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @NotNull
+    private static Map<String, List<String>> replicasByDataCenter(Map<String, String> hostToDatacenter,
+                                                                  Collection<String> replicas)
+    {
+        return replicas.stream().collect(Collectors.groupingBy(hostToDatacenter::get,
+                                                               Collectors.filtering(replicas::contains, toList())));
```

Comment:
The `adapter` subproject declare source compatibility of 1.8, but `Collectors.filtering` is available after java 9.

Reply:
Addressed